### ISSUE TITLE
Add some encryption and decryption to enhance the secure mechanism

### ISSUE THREAD

**Issue by lucascheng24** [lucascheng24/COMP3334-Discussion-forum#1]
any idea?

what content to be encrypted. and how to decrypt it?

**Comment by bardzjhe** [comment]
@kelvin0511 @gzlzgzl usually has phenomenal ideas so please suggest them here. just wanna remind you that, as mentioned by Dennis, it's not the more secure the better. Effectiveness matters as well. Sorry just confused two Kevins in our group.

**Comment by gzlzgzl** [comment]
Since posts on the discussion forum are visible to all users, we only need encryption between the client and the server. We can use RSA to achive this (available in the cryptico package), and here is a demo found on https://blog.csdn.net/junxuezheng/article/details/109824552：

const cryptico = require('cryptico')
// The passphrase used to repeatably generate this RSA key.
var PassPhrase = "The Moon is a Harsh Mistress."; 
// The length of the RSA key, in bits.
var Bits = 1024; 
var date1 = new Date().getTime()
var MattsRSAkey = cryptico.generateRSAKey(PassPhrase, Bits);
console.log(new Date().getTime()-date1) //
// Matt's public key string can then be generated like this:
var MattsPublicKeyString = cryptico.publicKeyString(MattsRSAkey);
// var publicKey =keys.exportKey('pkcs8-public-pem')
// var privateKey =keys.exportKey('pkcs1-private-pem')
console.log(MattsPublicKeyString.length)  //172，1024Bits对应的密钥长度为172
// 需要加密的信息PlainText
var PlainText = "Matt, I need you to help me with my Starcraft strategy.";
// Encrypting a message 加密
var EncryptionResult = cryptico.encrypt(PlainText, MattsPublicKeyString);
// Decrypting a message 解密
var DecryptionResult = cryptico.decrypt(EncryptionResult.cipher, MattsRSAkey);
console.log(DecryptionResult.plaintext)

Basically, the client shares a public key to the server when he registers, and keeps the private key on the local machine. The server records the public key permanently in the DB. At the same time, the server generates a key pair for each client, and sends the public key to the client while storing the private key in DB. So the client and server both keep each other's public key as well as their own private key.

When the client sends a message to the server, he encrypts the message body with the server's public key, and the server decrypts using the private key. It's the same the other way around.

**Comment by gzlzgzl** [comment]
For the first time of registering, we should probably add an invitation code (organization password).  This is because the project is designed to serve a small community. If anyone can register, then any encryption becomes meaningless. The hacker only needs to register an account if he wants to see all the post contents. This password can be set as a constant for our project, if we only want to open one community. In theory, we can have one community per password.

To avoid this password being leaked through man-in-the-middle attacks, it is necessary to deploy a password protocol. An example is the Encryption-with-Password + PKC：(lecture 6 slide 14)
![image](https://user-images.githubusercontent.com/93870892/232325011-d3b7ee92-0e4a-409f-811a-c1d7b27a959d.png)
Before registering, the client should know the correct invitation code pwd. Then W is calculated as SHA256(pwd). To register, the client sends the credentials (user name etc.) together with his public key, as mentioned before. Then, the server generates a challenge R and encrypts R with that public key, and PKE(R) is sent to the client. Upon receiving that, the client uses his private key to get R, and encrypts R with W using polyalphabetic cipher (lecture 2 slide 35). Then he sends EncW(R) to the server. The server decrypts and checks if it equals R. If it does, then the client is authenticated and is formally registered. Otherwise, he is not allowed to register.

**Comment by gzlzgzl** [comment]
Note that the client does not need to authenticate the server. Once joining the community, he can see all the posts and members in it and will automatically recognize whether its identity is true or not. Therefore, the EKE protocal need not to be adopted following the principle of adequate protection.
